File: src/domain/models/ripple_review.py
"""RippleReview aggregate root for the Ticket Freshness bounded context.

When a ticket closes, a RippleReview is created to flag open dependents
and siblings that may be affected by the change.

Invariants:
1. Non-empty context diff (enforced by ContextDiff VO __post_init__).
2. Only open tickets can be flagged.
3. Stacking allowed — multiple tickets can be flagged per review.
4. Explicit review required to clear a flag (ticket must be flagged first).
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from src.domain.events.ticket_freshness_events import FlagCleared, TicketFlagged
from src.domain.models.errors import InvariantViolationError

if TYPE_CHECKING:
    from src.domain.models.ticket_freshness import ContextDiff


class RippleReview:
    """Aggregate root: manages flagging and clearing of freshness flags.

    Attributes:
        review_id: Unique identifier for this review.
        closed_ticket_id: The ticket whose closure triggered this review.
        context_diff: Description of what the closed ticket produced.
    """

    def __init__(
        self,
        review_id: str,
        closed_ticket_id: str,
        context_diff: ContextDiff,
    ) -> None:
        self.review_id = review_id
        self.closed_ticket_id = closed_ticket_id
        self.context_diff = context_diff
        self._flagged_ticket_ids: list[str] = []
        self._events: list[TicketFlagged | FlagCleared] = []

    # -- Properties -----------------------------------------------------------

    @property
    def flagged_tickets(self) -> tuple[str, ...]:
        """Currently flagged ticket IDs (defensive copy)."""
        return tuple(self._flagged_ticket_ids)

    @property
    def events(self) -> tuple[TicketFlagged | FlagCleared, ...]:
        """Domain events produced by this aggregate (defensive copy)."""
        return tuple(self._events)

    # -- Commands -------------------------------------------------------------

    def flag_ticket(
        self, ticket_id: str, *, is_open: bool, flagged_at: str = ""
    ) -> None:
        """Flag a ticket for review.

        Args:
            ticket_id: The ticket to flag.
            is_open: Whether the ticket is currently open.
            flagged_at: ISO timestamp when the flag was created.

        Raises:
            InvariantViolationError: If the ticket is not open (invariant 2).
        """
        if not is_open:
            msg = f"Only open tickets can be flagged; '{ticket_id}' is not open"
            raise InvariantViolationError(msg)

        self._flagged_ticket_ids.append(ticket_id)
        self._events.append(
            TicketFlagged(
                review_id=self.review_id,
                ticket_id=ticket_id,
                context_diff=self.context_diff,
                flagged_at=flagged_at,
            )
        )

    def clear_flag(self, ticket_id: str, *, cleared_at: str = "") -> None:
        """Clear a freshness flag after explicit review.

        Args:
            ticket_id: The ticket whose flag to clear.
            cleared_at: ISO timestamp when the flag was cleared.

        Raises:
            InvariantViolationError: If the ticket is not flagged (invariant 4).
        """
        if ticket_id not in self._flagged_ticket_ids:
            msg = f"Ticket '{ticket_id}' is not flagged and cannot be cleared"
            raise InvariantViolationError(msg)

        self._flagged_ticket_ids.remove(ticket_id)
        self._events.append(
            FlagCleared(
                review_id=self.review_id,
                ticket_id=ticket_id,
                cleared_at=cleared_at,
            )
        )

File: internal/ticket/domain/ripple_review.go
package domain

import (
	"fmt"

	domainerrors "github.com/alty-cli/alty/internal/shared/domain/errors"
)

// ReviewChecklistTemplate is the standard review checklist for ripple reviews.
const ReviewChecklistTemplate = `**Review checklist:**
- [ ] Read the description -- does it still match the new context?
- [ ] Check acceptance criteria -- are any obsolete, incomplete, or contradicted?
- [ ] Verify DDD alignment -- do bounded-context boundaries still hold?
- [ ] Update or dismiss -- apply changes, or mark as unchanged if still valid.`

// RippleReview is the aggregate root for managing freshness flags.
type RippleReview struct {
	reviewID         string
	closedTicketID   string
	contextDiff      ContextDiff
	flaggedTicketIDs []string
	events           []any
	completed        bool
}

// NewRippleReview creates a RippleReview aggregate root.
func NewRippleReview(reviewID, closedTicketID string, contextDiff ContextDiff) *RippleReview {
	return &RippleReview{
		reviewID:       reviewID,
		closedTicketID: closedTicketID,
		contextDiff:    contextDiff,
	}
}

// ReviewID returns the review identifier.
func (r *RippleReview) ReviewID() string { return r.reviewID }

// ClosedTicketID returns the ID of the ticket whose closure triggered this review.
func (r *RippleReview) ClosedTicketID() string { return r.closedTicketID }

// ContextDiff returns the context diff describing what changed.
func (r *RippleReview) ContextDiff() ContextDiff { return r.contextDiff }

// FlaggedTickets returns a defensive copy of currently flagged ticket IDs.
func (r *RippleReview) FlaggedTickets() []string {
	out := make([]string, len(r.flaggedTicketIDs))
	copy(out, r.flaggedTicketIDs)
	return out
}

// Events returns a defensive copy of domain events.
func (r *RippleReview) Events() []any {
	out := make([]any, len(r.events))
	copy(out, r.events)
	return out
}

// FlagTicket flags a ticket for review. Only open tickets can be flagged.
func (r *RippleReview) FlagTicket(ticketID string, isOpen bool, flaggedAt string) error {
	if !isOpen {
		return fmt.Errorf("only open tickets can be flagged; '%s' is not open: %w",
			ticketID, domainerrors.ErrInvariantViolation)
	}
	r.flaggedTicketIDs = append(r.flaggedTicketIDs, ticketID)
	r.events = append(r.events, NewTicketFlagged(r.reviewID, ticketID, r.contextDiff, flaggedAt))
	return nil
}

// BuildRippleComment builds a structured ripple review comment.
func (r *RippleReview) BuildRippleComment() string {
	return fmt.Sprintf(
		"**Ripple review needed** -- `%s` was closed.\n\n**What changed:** %s\n\n%s",
		r.closedTicketID,
		r.contextDiff.Summary(),
		ReviewChecklistTemplate,
	)
}

// ClearFlag clears a freshness flag after explicit review.
func (r *RippleReview) ClearFlag(ticketID string, clearedAt string) error {
	for i, id := range r.flaggedTicketIDs {
		if id == ticketID {
			r.flaggedTicketIDs = append(r.flaggedTicketIDs[:i], r.flaggedTicketIDs[i+1:]...)
			r.events = append(r.events, NewFlagCleared(r.reviewID, ticketID, clearedAt))
			return nil
		}
	}
	return fmt.Errorf("ticket '%s' is not flagged and cannot be cleared: %w",
		ticketID, domainerrors.ErrInvariantViolation)
}

// Complete finalizes the ripple review and emits RippleReviewCreated event.
// Can only be called once per review.
func (r *RippleReview) Complete() error {
	if r.completed {
		return fmt.Errorf("review '%s' is already completed: %w",
			r.reviewID, domainerrors.ErrInvariantViolation)
	}
	r.completed = true
	event, err := NewRippleReviewCreated(r.reviewID, r.closedTicketID, len(r.flaggedTicketIDs))
	if err != nil {
		return fmt.Errorf("creating RippleReviewCreated event: %w", err)
	}
	r.events = append(r.events, event)
	return nil
}
